theano/gof/vm.py

```diff
@@ -889,6 +896,87 @@ def make_all(self, profiler=None, input_storage=None,
             compute_map[k] = [k.owner is None]
 
         thunks = []
+
+        # Collect Reallocation Info
+        compute_map_re = defaultdict(lambda: [0])
+        for var in fgraph.inputs:
+            compute_map_re[var][0] = 1
+
+        viewed_by = {}
+        for var in fgraph.variables:
+            viewed_by[var] = []
+        view_of = {}
+        reallocated_info = {}
+        dependencies = getattr(fgraph.profile, 'dependencies', {})
+        pre_allocated = set([])
+
+        for idx in range(len(order)):
+            node = order[idx]
+            dmap = getattr(node.op, 'destroy_map', None)
+            vmap = getattr(node.op, 'view_map', None)
+
+            idx_o = 0
+            for out in node.outputs:
+                for var in node.outputs:
+                    compute_map_re[var][0] = 1
+                ins = None
+                if dmap and idx_o in dmap:
+                    idx_v = dmap[idx_o]
+                    assert len(
+                        idx_v) == 1, "Here we only support the possibility to destroy one input"
+                    ins = node.inputs[idx_v[0]]
+                if vmap and idx_o in vmap:
+                    assert ins is None
+                    idx_v = vmap[idx_o]
+                    assert len(
+                        idx_v) == 1, "Here we only support the possibility to view one input"
+                    ins = node.inputs[idx_v[0]]
+                if ins is not None:
+                    assert isinstance(ins, theano.Variable)
+                    origin = view_of.get(ins, ins)
+                    view_of[out] = origin
+                    viewed_by[origin].append(out)
+                idx_o += 1
+
+            for ins in node.inputs:
+                assert not (ins in view_of and viewed_by[ins])
+                if (getattr(ins, 'ndim', None) == 0 and not storage_map[ins][0]
+                        and ins not in fgraph.outputs and ins.owner
+                        and all([compute_map_re[v][0] for v in dependencies.get(ins, [])])):
+                    # Constant Memory cannot be changed, Constant storage_map
+                    # has a value here
+                    reuse_out = None
+                    if ins not in view_of and not viewed_by.get(ins, []):
+                        # where gc
+                        for i in range(idx + 1, len(order)):
+                            if reuse_out:
+                                break
+                            for out in order[i].outputs:
+                                if getattr(out, 'ndim', None) == 0 and out not in pre_allocated:
```

Comment:
You need to add an extra condition. I think this is what make tests fail.

add out.type == ins.type